### DISCUSSION TITLE
what's the proper way to create global interceptors using $fetch

### DISCUSSION THREAD

**Discussion by saifobeidat** [nuxt/nuxt#22187] [Questions]
Hello

I was searching for a way to create a general interceptor in the Nuxt docs and in ofetch docs as well, but I didn't find anything..

there was a hint in the ofetch docs that it can be used like this:

```
export default defineNuxtPlugin(async nuxtApp => {
  $fetch.create({
    async onResponse(res) {
      console.log(re0s
    },
  });
 })
```

But the interceptor doesn't get called at all..

and then i tried this from stackoverflow:

```
export default defineNuxtPlugin(async nuxtApp => {
   globalThis.$fetch = ofetch.create({
    async onResponse(res) {
      console.log(res)
    },
  });
 })
```

and it works fine after each request it will be called... but i am not sure if this a good way or not as this was not documented anywhere

**Comment by angelhdzmultimedia** [comment]
`plugins/interceptors/auth/index.ts`

```ts
import { NuxtApp } from 'nuxt/app'

export type FetchClient = NuxtApp & {
  $client?: typeof $fetch
}

export default defineNuxtPlugin(async nuxtApp => {
  const client = $fetch.create<string>({
    async onResponse(res) {
      console.log(res)
      console.log('Interceptor working!')
    },
  });

  return {
    provide: {
      client
    }
  }
 })
```

`pages/index.vue`

```html
<script setup lang="ts">
import { FetchClient } from 'plugins/interceptors/auth'

const {$client}: FetchClient = useNuxtApp()

async function load(): Promise<void> {
  const data = await $client!('/api/auth/login', {
    method: 'post'
  })
  Notify.create({
    message: data ?? ''
  })
}
</script>

<template>
```

![image](https://github.com/nuxt/nuxt/assets/10191085/6a3b6223-308a-4e40-ae03-5d4e46bde50f)

**Comment by Daisigu** [comment]
You can use nuxt-advanced-fetch module
https://www.npmjs.com/package/nuxt-advanced-fetch
